src/connectors/huggingface/huggingface_dataset_connector.py

```diff
@@ -0,0 +1,143 @@
+        params = {"dataset": dataset_name}
+        error_msg = "Error while fetching splits from HuggingFace"
+        response_json = HuggingFaceDatasetConnector._get(url, error_msg, params=params)
+        split_infos = [
+            file
+            for file in response_json["splits"]
+            if file["config"] == config and file["split"] == split
+        ]
+        if len(split_infos) != 1:
+            msg = (
+                f"HuggingFace's split endpoint does not contain {config=}, {split=} for "
+                f"dataset {dataset_name} (or returns multiple)."
+            )
+            raise HTTPException(status_code=404, detail=msg)
+        split_info = split_infos[0]
+
+        url = "https://datasets-server.huggingface.co/parquet"
+        params = {"dataset": dataset_name}
+        error_msg = "Error while fetching parquet data from HuggingFace"
+        response_json = HuggingFaceDatasetConnector._get(url, error_msg, params=params)
+        file_infos = [
+            file
+            for file in response_json["parquet_files"]
+            if file["config"] == config and file["split"] == split
+        ]
+        if len(file_infos) != 1:
+            msg = (
+                f"HuggingFace's parquet endpoint does not contain {config=}, {split=} for "
+                f"dataset {dataset_name} (or returns multiple)."
+            )
+            raise HTTPException(status_code=404, detail=msg)
+        file_info = file_infos[0]
+
+        # TODO: decide our output format for datasets.
+        #  If we want extra information, e.g. the number of features, this works:
+        # url = "https://datasets-server.huggingface.co/first-rows"
+        # params = {"dataset": dataset_name, "config": config, "split": split}
+        # error_msg = "Error while fetching first-rows from HuggingFace"
+        # response_json = HuggingFaceDatasetConnector._get(url, error_msg, params=params)
+        # n_features = len(response_json["features"])
+
+        return Dataset(
+            name=dataset.name,
+            identifier=dataset.platform_specific_identifier,
+            distribution=DataDownload(contentUrl=file_info["url"], encodingFormat="parquet"),
+            size=QuantitativeValue(value=split_info["num_examples"]),
+            isAccessibleForFree=True,
+            includedInDataCatalog=DataCatalog(name="HuggingFace"),
+        )
+
+    def fetch_all(self) -> list[DatasetDescription]:
+        url = "https://datasets-server.huggingface.co/valid"
+        error_msg = "Error while fetching all data from HuggingFace"
+        response_json = HuggingFaceDatasetConnector._get(url, error_msg)
+        return list(self._yield_datasets(response_json["valid"]))
+
+    def _yield_datasets(self, dataset_names) -> typing.Iterator[DatasetDescription]:
+        """Yield a Dataset for each (name, config, split) combination."""
+        for dataset_name in dataset_names:
+            yield from self._yield_datasets_with_name(dataset_name)
```

Comment:
It's nice that you are yielding, given the work required to fetch each dataset. However, it seems like the only exposure right now is through `fetch_all` which turns it into a `list` anyway. Perhaps `fetch_all` should also return an iterator so the users can benefit from this lazy iteration.